File: GmApps/src/app/audit-exam/exam.service.ts
import { Observable } from "rxjs";

import { Injectable, Inject } from "@angular/core";
import { HttpClient } from "@angular/common/http";

import { Exam } from "./exam";
import { ExamSubmission } from "../models/ExamSubmission";
import { ExamResult } from "../models/ExamResult";
import { map } from "rxjs/operators";

@Injectable()
export class ExamService {

    constructor(private http: HttpClient) {
    }

    currentExamVersions(): string[] {
        return ["201705", "201707"];
    }

    getExam(version: string): Observable<Exam> {
        return this.http.get<Exam>(`/assets/audit-exam-${version}.json`);
    }

    saveExamDraft(draft: ExamSubmission): Observable<boolean> {
        return this.http.put("/api/AuditExam/Draft", draft, { withCredentials: true, observe: "response" })
            .pipe(map(resp => resp.ok));
    }

    getExamDraft(version: string): Observable<ExamSubmission> {
        return this.http.get<ExamSubmission>("/api/AuditExam/Draft", { params: { version: version }, withCredentials: true });
    }

    submitAnswer(submission: ExamSubmission): Observable<ExamResult> {
        return this.http.post<ExamResult>("/api/AuditExam/Submit", submission, { withCredentials: true });
    }

    getExamResult(version: string): Observable<ExamResult|null> {
        return this.http.get<ExamResult>("/api/AuditExam/Result", { params: { version: version }, withCredentials: true });
    }

    getAllExamResult(): Observable<ExamResult[] | null> {
        return this.http.get<ExamResult[]>("/api/AuditExam/Results", { withCredentials: true });
    }

    // Admin Only.
    getExamResultForUser(user: string, version: string): Observable<ExamResult> {
        return this.http.get<ExamResult>("/api/AuditExam/ResultForUser", { params: { version: version, user: user }, withCredentials: true });
    }
}

File: GmGard.Client/Services/ExamService.cs
using System.Net.Http.Json;
using GmGard.Client.Models;

namespace GmGard.Client.Services;

public class ExamService
{
    private readonly HttpClient _http;
    public ExamService(HttpClient http) => _http = http;

    public string[] CurrentExamVersions() => ["201705", "201707"];

    public async Task<Exam?> GetExamAsync(string version)
    {
        try { return await _http.GetFromJsonAsync<Exam>($"/assets/audit-exam-{version}.json"); }
        catch { return null; }
    }

    public async Task<ExamSubmission?> GetDraftAsync(string version)
    {
        try { return await _http.GetFromJsonAsync<ExamSubmission>(ApiRoutes.AuditExam.Draft(version)); }
        catch { return null; }
    }

    public async Task<bool> SaveDraftAsync(ExamSubmission draft)
    {
        try { var r = await _http.PutAsJsonAsync(ApiRoutes.AuditExam.DraftPut, draft); return r.IsSuccessStatusCode; }
        catch { return false; }
    }

    public async Task<ExamResult?> SubmitAsync(ExamSubmission sub)
    {
        try
        {
            var resp = await _http.PostAsJsonAsync(ApiRoutes.AuditExam.Submit, sub);
            if (!resp.IsSuccessStatusCode) return null;
            return await resp.Content.ReadFromJsonAsync<ExamResult>();
        }
        catch { return null; }
    }

    public async Task<ExamResult?> GetResultAsync(string version)
    {
        try { return await _http.GetFromJsonAsync<ExamResult>(ApiRoutes.AuditExam.Result(version)); }
        catch { return null; }
    }

    public async Task<List<ExamResult>?> GetAllResultsAsync()
    {
        try { return await _http.GetFromJsonAsync<List<ExamResult>>(ApiRoutes.AuditExam.Results); }
        catch { return null; }
    }

    public async Task<ExamResult?> GetResultForUserAsync(string user, string version)
    {
        try { return await _http.GetFromJsonAsync<ExamResult>(ApiRoutes.AuditExam.ResultForUser(version, user)); }
        catch { return null; }
    }
}
